trie/utils/db.py

```diff
@@ -0,0 +1,63 @@
+import contextlib
+
+MISSING = object()
+
+
+class ScratchDB:
+    """
+    A wrapper of basic DB objects with uncommitted DB changes stored in local cache,
+    which represents as a dictionary of database keys and values.
+    This class should be usable as a context manager, the changes either all fail or all succeed.
+    Upon exiting the context, it writes all of the key value pairs from the cache into
+    the underlying database. If anyerror occurred before committing phase,
+    we would not apply commits at all.
+    """
+    def __init__(self, wrapped_db):
+        self.wrapped_db = wrapped_db
+        self.cache = {}  # type: Dict[bytes, bytes]
+        self.deletes = []
+
+    #
+    # Dictionary API
+    #
+    # if not key is found, return None
+    def __getitem__(self, key):
+        if key in self.cache:
+            return self.cache[key]
+        else:
+            try:
+                current_value = self.wrapped_db.get(key)
```

Comment:
I think this will return incorrectly if the key is in the delete set.  Seems like another check needs to be added to return `None` if the key is in the delete set.